identity_did/src/document/core_document.rs

```diff
@@ -142,273 +247,367 @@ where
-        self.key_agreement.append(MethodRef::Embed(method))
+        self.data.key_agreement.append(MethodRef::Embed(method))
       }
       MethodScope::VerificationRelationship(MethodRelationship::CapabilityDelegation) => {
-        self.capability_delegation.append(MethodRef::Embed(method))
+        self.data.capability_delegation.append(MethodRef::Embed(method))
       }
       MethodScope::VerificationRelationship(MethodRelationship::CapabilityInvocation) => {
-        self.capability_invocation.append(MethodRef::Embed(method))
+        self.data.capability_invocation.append(MethodRef::Embed(method))
       }
     };
 
     Ok(())
   }
 
-  /// Removes all references to the specified [`VerificationMethod`].
-  ///
-  /// # Errors
+  /// Removes and returns the [`VerificationMethod`] from the document.
   ///
-  /// Returns an error if the method does not exist.
-  pub fn remove_method(&mut self, did: &DIDUrl<D>) -> Result<()> {
-    let was_removed: bool = [
-      self.authentication.remove(did),
-      self.assertion_method.remove(did),
-      self.key_agreement.remove(did),
-      self.capability_delegation.remove(did),
-      self.capability_invocation.remove(did),
-      self.verification_method.remove(did),
+  /// # Note
+  /// All references to the method found in the document will be removed (also in the case where the method is not
+  /// contained in this document).
+  pub fn remove_method(&mut self, did: &DIDUrl<D>) -> Option<VerificationMethod<D, U>> {
+    for method_ref in [
+      self.data.authentication.remove(did),
+      self.data.assertion_method.remove(did),
+      self.data.key_agreement.remove(did),
+      self.data.capability_delegation.remove(did),
+      self.data.capability_invocation.remove(did),
     ]
-    .contains(&true);
-
-    if was_removed {
-      Ok(())
-    } else {
-      Err(Error::MethodNotFound)
+    .into_iter()
+    .flatten()
+    {
+      if let MethodRef::<D, U>::Embed(embedded_method) = method_ref {
+        // embedded methods cannot be referenced, or be in the set of general purpose verification methods hence the
+        // search is complete
+        return Some(embedded_method);
+      }
     }
+
+    self.data.verification_method.remove(did)
   }
 
+  /// Adds a new [`Service`] to the document.
+  ///
+  /// # Errors
+  ///
+  /// Returns an error if there already exists a service or (verification) method with the same identifier.
+  pub fn insert_service(&mut self, service: Service<D, V>) -> Result<()> {
+    let service_id = service.id();
+    let id_shared_with_method = self
+      .verification_relationships()
+      .map(|method_ref| method_ref.id())
+      .chain(self.verification_method().iter().map(|method| method.id()))
+      .any(|id| id == service_id);
+
+    ((!id_shared_with_method) && self.service_mut_unchecked().append(service))
+      .then_some(())
+      .ok_or(Error::InvalidServiceInsertion)
+  }
```

Comment:
Your suggestion would be the way to go if we decide to disallow referencing verification methods in other DID documents, but as long as we allow that `self.resolve_method(service.id(), None).is_none()` will return false positives, which is why I wrote it the way it is. The need for this complexity definitely warrants a comment however so I will add that.

Reply:
I see. How about this then?

```rust
if self.resolve_method(service.id(), None).is_some() || !self.service_mut_unchecked().append(service) {
  Err(Error::InvalidServiceInsertion)
} else {
  Ok(())
}
```

Reply:
If I have a DID document of the form: 
```
{
      "id": "did:example:1234",
      "capabilityDelegation": [
        "did:other-method:5678#key1"
      ]
}
```
and I call `insert_service` with a service whose `id` is `did:other-method:5678#key1` then the call to `self.resolve_method(service.id(), None)` returns `None` because the return type of `resolve_method` is `VerificationMethod` and not `MethodRef`.